### DISCUSSION TITLE
Ploting on a canvas / PDF document

### DISCUSSION THREAD

**Discussion by zell-mbc** [patrykandpatrick/vico#668] [Questions] [answered]
I am publishing my own little Android app which allows user to track certain data, create the associated charts as well as creating PDF documents to share their data with others.

I recently converted the app over to Compose, with the exception of the charts which are still view based.
For the charts I currently use AndroidPlot which is working ok. What I am struggling with though is to plot the same charts onto a PDF canvas (android.graphics.pdf.PdfDocument).

The PDF issue and the fact that AndroidPlot seems to be abandoned has me looking for alternative libraries for quite some time.

Therefore my question, does Vico support plotting on a canvas/pdf object, are there any plans to support this scenario?
Maybe using the compose API for on-screen and the view API  -> convert to bitmap for the PDF?

https://f-droid.org/en/packages/com.zell_mbc.medilog/

Edit:
Went ahead with Vico no matter the answer to my question above and have to say that I am impressed!
I don't think I have even started to grasp how things hang together, but I managed to convert my most complicated chart over to Vico and the result looks really beautiful!
Now the canvas question is even more important to me because I really want to move to Vico now. SHould I raise this as an issue rather than a discussion?

**Comment by patrickmichalik** [comment]
Hello, and thanks for the kind words! Regarding the choice between an issue and a discussion, we use discussions for questions, so you’re in the right place. I’ve moved the discussion to the “Q&A” category.

While Vico doesn’t have a dedicated API for image generation, what you’re asking about is certainly possible. You can create images of both views and composables, with the former being significantly simpler when you need an image of something that’s not shown in the app’s UI. Am I correct in assuming that a dedicated chart is required for the PDF? It would also be possible to capture an image of the chart in the UI, but the scaling of the chart in the PDF would then be device-dependent.

**Reply by zell-mbc** [reply]
You assumed correctly, what I am after is a dedicated chart for the PDF document, sized for an A4 or letter document and compiled in a headless/invisible way.
The content of the PDF is based on the raw data in a nicely formatted way, a summary page showing statistics and now I'd like to add a chart to round things up.

I wouldn't go for a solution which requires capturing a UI image. For the scaling reasons you describe, but even more so because the workflow would be highly confusing for the user, because I build the PDF in a separate co-routine and pass it on to the devices PDF viewer or other apps once done. It would be very hard to explain why all of a sudden a chart would pop up for a second out of nowhere just to be able to grab the screen.

**Reply by Gowsky** [reply, marked as answer]
> even more so because the workflow would be highly confusing for the user, because I build the PDF in a separate co-routine and pass it on to the devices PDF viewer or other apps once done. It would be very hard to explain why all of a sudden a chart would pop up for a second out of nowhere just to be able to grab the screen.

We were under the impression that a chart was already shown on screen, making it possible to capture without disruption. That's what @patrickmichalik was referring to, presenting it as the inferior opposite of having a dedicated chart. He wasn't talking about having a chart appear for a split second just to be captured.

This follows from the scaling comment — those concerns apply only to a normal chart that's in the UI regardless of capturing, for the user to view. A chart that pops up just to be captured could just as well be sized for the PDF, so there wouldn't be any scaling issues.

> You assumed correctly

Then the simplest solution is to use views, as @patrickmichalik mentioned. Please see [here](https://stackoverflow.com/questions/17189809/is-it-possible-to-take-a-screenshot-of-a-view-without-showing-the-view) and [here](https://patrykandpatrick.com/vico/wiki/cartesian-charts/data#manual-cartesianchartmodel-creation).

**Reply by zell-mbc** [reply]
Thanks for getting back. That's a bit of a bummer because it means I essentially have to create the chart logic twice…or maybe I migrate the UI compose charts to views. I'll have a look and decide what's easier.

**Reply by Gowsky** [reply]
You're welcome. You can try using Compose for this, but as per @patrickmichalik's comment, capturing composables without displaying them is relatively complicated.
